app/src/main/java/com/x8bit/bitwarden/data/platform/manager/BiometricsEncryptionManagerImpl.kt

```diff
@@ -88,16 +92,21 @@ class BiometricsEncryptionManagerImpl(
     }
 
     /**
-     * Generates a [SecretKey] from which the [Cipher] will be generated.
+     * Generates a [SecretKey] from which the [Cipher] will be generated, or `null` if a key cannot
+     * be generated.
      */
-    private fun generateKey(): SecretKey {
+    private fun generateKeyOrNull(): SecretKey? {
         val keyGen = KeyGenerator.getInstance(
             KeyProperties.KEY_ALGORITHM_AES,
             ENCRYPTION_KEYSTORE_NAME,
         )
         keyGen.init(keyGenParameterSpec)
-        keyGen.generateKey()
-        return requireNotNull(getSecretKey())
+        try {
+            keyGen.generateKey()
+        } catch (e: ProviderException) {
+            return null
+        }
+        return getSecretKey()
```

Comment:
Can this throw as well?

Reply:
Yup. It's masking several exceptions that could be thrown by `KeyGenerator` and `keystore` calls. Am I correct in thinking we need to set `settingsDiskSource.systemBiometricIntegritySource = null` for all of these exceptions?

Reply:
I think what Brian wants is for any function that can throw an exception to be updated to handle the exception and return null.

Like:
```kotlin
    private fun getSecretKeyOrNull(): SecretKey? {
        keystore.load(null)
        return try {
            keystore.getKey(ENCRYPTION_KEY_NAME, null) as? SecretKey
        } catch (blah) {
            null
        }
    }
```

And then the consumers if those APIs would set things to null based on cirumstance.

```kotlin
    override fun getOrCreateCipher(userId: String): Cipher? {
        val secretKey = getSecretKeyOrNull()
            ?: generateKeyOrNull()
            ?: run {
                // user removed all biometrics from the device
                settingsDiskSource.systemBiometricIntegritySource = null
                return null
            }
```

Reply:
Yeah exactly...that's what I'm thinking. Any function call that can throw gets wrapped and then the places that call those functions can decide what to do in the failure case. Another option is to use `runCatching` in order to return a `Result` and then use `fold` or any of the other tools we need.